common/src/autogluon/common/utils/deprecated.py

```diff
@@ -1,26 +1,230 @@
-import logging
-import warnings
 import functools
-from typing import Optional
+import inspect
+import warnings
+from typing import Any, Dict, Optional
+
+from packaging import version
 
+from ..version import __version__
 
-logger = logging.getLogger(__name__)
 
+def _deprecation_warning(
+    old: str,
+    new: Optional[str] = None,
+    custom_warning_msg: Optional[str] = None,
+    version_to_remove: Optional[str] = None,
+    error: bool = False,
+):
+    msg = custom_warning_msg
+    if msg is None:
+        msg = f"`{old}` has been deprecated. "
+        msg += f"And will be removed in version {version_to_remove} " if version_to_remove is not None else ""
+        msg += f"Please use `{new}` instead" if new is not None else ""
 
-def deprecated(message: str = "", version_removed: Optional[str] = None, *args, **kwargs):
-    def deprecated_decorator(func):
-        warning_message = (
-            f"Calling deprecated function {func.__name__}. This function is deprecated "
-            + "and will be removed in "
-            + (f"v{version_removed}. " if version_removed else "a future release. ")
-            + message
+    if error:
+        raise ValueError(msg)
```

Comment:
why ValueError? Is this the commonly used error for deprecation errors?

Reply:
There's no built-in deprecation error/exception. So just used the general ValueError. For the warning, we are throwing deprecation warning

Reply:
ok sounds good